wandb/jupyter.py

```diff
@@ -21,69 +24,92 @@
-        return self.displayed
 
-    def _repr_html_(self) -> str:
-        try:
-            self.displayed = True
+class _WandbCellMagicState:
+    """State for a cell with the %%wandb cell magic."""
 
-            # Display the explicit resource if specified.
-            if isinstance(self.path, str):
-                return self.api.from_path(self.path).to_html(
-                    self.height,
-                    hidden=False,
-                )
+    def __init__(self, *, height: int) -> None:
+        """Initializes the %%wandb cell magic state.
 
-            # Otherwise, display the global run, if there is one.
-            elif wandb.run:
-                return wandb.run.to_html(self.height, hidden=False)
+        Args:
+            height: The desired height for displayed iframes.
+        """
+        self._height = height
+        self._already_displayed = False
 
-            # Otherwise, display the user's project.
-            elif wandb.Api().api_key is None:
-                return (
-                    "You must be logged in to display wandb in Jupyter."
-                    " Please run `wandb.login()`."
-                )
-            else:
-                return self.api.project(
-                    "/".join(
-                        [
-                            wandb.Api().default_entity,
-                            wandb.util.auto_project_name(None),
-                        ]
-                    )
-                ).to_html(self.height, hidden=False)
+    def display_if_allowed(self, run: wandb_run.Run) -> None:
+        """Display a run's iframe if one is not already displayed.
 
-        except Exception as e:
-            return f"Error displaying wandb object.<br/>{e}"
+        Args:
+            run: The run to display.
+        """
+        if self._already_displayed:
+            return
+        self._already_displayed = True
+
+        _display_wandb_run(run, height=self._height)
+
+
+_current_cell_wandb_magic: _WandbCellMagicState | None = None
+
+
+def _display_by_wandb_path(path: str, *, height: int) -> None:
+    """Display a wandb object (usually in an iframe) given its URI.
+
+    Args:
+        path: A path to a run, sweep, project, report, etc.
+        height: Height of the iframe in pixels.
+    """
+    api = wandb.Api()
+
+    try:
+        obj = api.from_path(path)
+
+        IPython.display.display_html(
+            obj.to_html(height=height),
+            raw=True,
+        )
+    except wandb.Error:
+        traceback.print_exc()
+        IPython.display.display_html(
+            f"Path {path!r} does not refer to a W&B object you can access.",
+            raw=True,
+        )
+
+
+def _display_wandb_run(run: wandb_run.Run, *, height: int) -> None:
+    """Display a run (usually in an iframe).
+
+    Args:
+        run: The run to display.
+        height: Height of the iframe in pixels.
+    """
+    IPython.display.display_html(
+        run.to_html(height=height),
```

Comment:
not for this pr, but do you think we can remove this from the run's methods? is it needed outside of this usage?

Reply:
Yeah I have a feeling these methods only exist for our internal Jupyter stuff, so we should remove them. I'll get to that refactor eventually.